### PR TITLE
[luci/service] Migrate Quantize shape inference rule to sinf::Algorithm

### PR DESCRIPTION
**Pull request by kyeong8139**
When testing dynamic shape inference, the quantize operation successfully handles it.
I have added tests for this case and migrated the inference code.

Related to: https://github.com/Samsung/ONE/issues/13697
draft: https://github.com/Samsung/ONE/pull/13779

> ONE-DCO-1.0-Signed-off-by: kyeong8139 [kyeong8139@gmail.com](mailto:kyeong8139@gmail.com)

### CODE REVIEW

**Comment by seanshpark** [comment]
> add dynamic shape inference test for Quantize

I was expecting to see test codes... but there exist code migration, but no specific algorithm changes for dynamic shape support.

Can you update the title to express this?

@zetwhite , @shs-park , Q) do we MUST relocate codes ? I don't fully understand/agree on existing design that @llFreetimell made.

**Comment by shs-park** [comment]
> Q) do we MUST relocate codes ? I don't fully understand/agree on existing design that @llFreetimell made.

I think this is not a MUST todo,
but we need to have some consistency of the location for the shape inference when we(with SSAFY members)'re working on this time.

The Ops we want to implement this time were discussed in [this issue](https://github.com/Samsung/ONE/issues/13730) and we have plan to implement them in the location of @llFreetimell's firstly.

**Comment by seanshpark** [comment]
> but we need to have some consistency of the...

Q) does `Quantize` have any problem with dynamic shape inference?

Thoughts,
- we can relocate `Quantize` with additional tests with dynamic shape inference
- we can skip this `Quantize` if there is no problem as of now. eventually it looks like migration is needed. that is if there exist other Ops that need to migrate for dynamic shape infer, lets work on that ops with higher priority.
- current title/comments needs update to express for actual code changes.

**Comment by zetwhite** [comment]
I think re-location is helpful to distinguish which op is ready to support dynamic shape and which op is not yet. If not, others might try to re-investigate to support dynamic shapes for the Quantize node. 

> we can relocate Quantize with additional tests with dynamic shape inference 

For me, this option looks good. 

> current title/comments needs update to express for actual code changes. 

For PR title, how about 'Migrate Quantize shape inference rule to sinf::Algorithm'?

**Comment by seanshpark** [comment]
> ... is helpful to distinguish which op is ready to ...

👍🏻 

>  'Migrate Quantize shape inference rule to sinf::Algorithm'?

good to me :)

**Comment by shs-park** [comment]
> Q) does `Quantize` have any problem with dynamic shape inference?

AFAIK roughly, this Op is related to the `prompt parsing` and `token gen` parts, from https://github.com/Samsung/ONE/issues/13697.

@jinevening, could you explain a little bit more..?

**Comment by kyeong8139** [comment]
> For PR title, how about 'Migrate Quantize shape inference rule to sinf::Algorithm'?
I change the title to 'Migrate Quantize shape inference rule to sinf::Algorithm'. I think It is appropriate title about this, too.

**Comment by seanshpark** [comment]
From https://github.com/Samsung/ONE/pull/13794/commits

commit message is 

>  [luci-quantize] add dynamic shape tests
>
> This commit add dynamic shape inference tests for quantize.


can you please fix your commit?
github UI title/commit message won't update the commit itself.

**Comment by kyeong8139** [comment]
> github UI title/commit message won't update the commit itself.

I fixed my commit. 
I'd like to ask something. I changed [compiler/luci/service/src/Nodes/CircleQuantize.test.cpp](https://github.com/Samsung/ONE/pull/13794/commits/f559b83e59e4d4de3d6d839fb12513b3895a10bd#diff-b2e0f94566faab689d7247e9004d90561a007cf42b023bb31ffb5d37c672a0fe) to ensure that original quantize algorithm support dynamic shape inference. 
it isn't necessart if op don't have problem for supporting dynamic shape? or It isn't enough to test it?

**Comment by zetwhite** [comment]
> I changed [compiler/luci/service/src/Nodes/CircleQuantize.test.cpp](https://github.com/Samsung/ONE/pull/13794/files) to ensure that original quantize algorithm support dynamic shape inference.

I agree with this comment. I think adding tc to show it supports dynamic shape well is good idea.

**Comment by seanshpark** [comment]
> I fixed my commit.

https://github.com/Samsung/ONE/pull/13794/commits will show lots of commits
that will be accumulated (we use squash merge) when landing.

you need to update the first commit.

**Comment by jinevening** [comment]
> @jinevening, could you explain a little bit more..?

Quantize Op is necessary for mixed-precision LLM.

**Review by shs-park — DISMISSED** [review]
LGTM!
=)

File: compiler/luci/service/src/Nodes/CircleQuantize.cpp [hunk 1]
```diff
@@ -24,4 +27,10 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleQuantize *)
   return _graph->nodes()->create<luci::CircleQuantize>();
 }
 
+loco::TensorShape sinf::Algorithm::visit(const luci::CircleQuantize *node)
+{
+  const auto input_shape = luci::shape_get(node->input()).as<loco::TensorShape>();
```
Comment (jinevening) [inline compiler/luci/service/src/Nodes/CircleQuantize.cpp:32]: Could you use `circle_shape` function? `shape_get` is a temporary function, which will be removed in the future.

`CircleShapeInferenceHelper.h`
```cpp
// NOTE Functions in this namespace will be removed after new inference
//      algorithms are fully implemented.

// This function is temporary function for deprecating loco::shape_get
loco::NodeShape shape_get(const loco::Node *node);
```

File: compiler/luci/service/src/Nodes/CircleQuantize.cpp [hunk 2]
```diff
@@ -24,4 +27,10 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleQuantize *)
   return _graph->nodes()->create<luci::CircleQuantize>();
 }
 
+loco::TensorShape sinf::Algorithm::visit(const luci::CircleQuantize *node)
+{
+  const auto input_shape = circle_shape(node);
```
Comment (jinevening) [inline compiler/luci/service/src/Nodes/CircleQuantize.cpp:32]: It has to be input's shape.

**Comment by kyeong8139** [comment]
there are some troblems while using git rebase.
after close this pr, I write new pull request.

**Pull request closed without merging**